File: app/src/main/java/com/yanrou/dawnisland/content/SeriesContentActivity.java
package com.yanrou.dawnisland.content;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.Window;
import android.view.WindowManager;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.drakeet.multitype.MultiTypeAdapter;
import com.scwang.smartrefresh.layout.SmartRefreshLayout;
import com.yanrou.dawnisland.FooterView;
import com.yanrou.dawnisland.FooterViewBinder;
import com.yanrou.dawnisland.R;
import com.yanrou.dawnisland.ReplyDialog;
import com.yanrou.dawnisland.SeriesRecyclerOnScrollListener;
import com.yanrou.dawnisland.database.SeriesData;
import com.yanrou.dawnisland.util.ReadableTime;

import org.jetbrains.annotations.NotNull;

/**
 * @author suche
 */
public class SeriesContentActivity extends AppCompatActivity {
    private static final String TAG = "SeriesContentActivity";
    RecyclerView recyclerView;
    Toolbar toolbar;
    ActionBar actionBar;

    String id;
    String forumName;

    SmartRefreshLayout smartRefreshLayout;

    ReplyDialog replyDialog = null;

    SeriesData seriesData;

    private MultiTypeAdapter multiTypeAdapter;

    private SeriesContentViewModel viewModel;

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.series_content_menu, menu);
        return true;
    }


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_series_content);
        //TODO 应该更早初始化该类
        ReadableTime.initialize(this);
        //获取控件
        initView();

        actionBar.setDisplayHomeAsUpEnabled(true);
        //状态栏透明
        makeStatusBarTran(getWindow());

        Intent intent = getIntent();
        id = intent.getStringExtra("id");
        //TODO 这里应该传入fid，然后通过DB类获取到板块名称，而不是直接传入板块名称
        forumName = intent.getStringExtra("forumTextView");
        Log.d(TAG, "onCreate: " + id);
        toolbar.setTitle("A岛 · " + forumName);
        toolbar.setSubtitle(">>No." + id + " · adnmb.com");

        viewModel = new ViewModelProvider(this).get(SeriesContentViewModel.class);

        multiTypeAdapter = new MultiTypeAdapter();
        
        multiTypeAdapter.register(ContentItem.class, new ContentViewBinder(SeriesContentActivity.this));

        multiTypeAdapter.register(FooterView.class, new FooterViewBinder());

        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.addOnScrollListener(new SeriesRecyclerOnScrollListener() {
            @Override
            public void onLoadMore() {
                viewModel.loadMore();
            }
        });

        recyclerView.setAdapter(multiTypeAdapter);
        smartRefreshLayout.setEnableAutoLoadMore(false);
        smartRefreshLayout.setOnRefreshListener(refreshLayout -> viewModel.refresh());
        //TODO 首次获取放在view model中
        //presenter.loadFirstPage();


    }

    private void makeStatusBarTran(@NotNull Window window) {
        //这一步要做，因为如果这两个flag没有清除的话下面没有生效
        window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS | WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION);
        //设置布局能够延伸到状态栏(StatusBar)和导航栏(NavigationBar)里面
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS);
        //设置状态栏(StatusBar)颜色透明
        window.setStatusBarColor(getResources().getColor(R.color.colorPrimary));
        //设置导航栏(NavigationBar)颜色透明
        //window.setNavigationBarColor(getResources().getColor(R.color.colorPrimary));
    }

    void initView() {
        recyclerView = findViewById(R.id.series_content_recycleview);
        toolbar = findViewById(R.id.cotent_toolbar);
        setSupportActionBar(toolbar);
        actionBar = getSupportActionBar();
        smartRefreshLayout = findViewById(R.id.smart_refresh);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            //左上角的箭头按钮的动作
            case android.R.id.home:
                finish();
                break;
            case R.id.jump_page:
                break;
            case R.id.reply:
                //防止启动多次
                if (replyDialog == null) {
                    replyDialog = new ReplyDialog();
                    Bundle bundle = new Bundle();
                    bundle.putString("seriesId", id);
                    replyDialog.setArguments(bundle);
                }
                replyDialog.show(getSupportFragmentManager(), "reply");
                break;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/yanrou/dawnisland/content/SeriesContentActivity.kt
package com.yanrou.dawnisland.content

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.Window
import android.view.WindowManager
import android.widget.Toast
import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.drakeet.multitype.MultiTypeAdapter
import com.lxj.xpopup.XPopup
import com.lxj.xpopup.interfaces.SimpleCallback
import com.scwang.smartrefresh.layout.SmartRefreshLayout
import com.yanrou.dawnisland.R
import com.yanrou.dawnisland.SeriesRecyclerOnScrollListener
import com.yanrou.dawnisland.constant.TYPE_KEY
import com.yanrou.dawnisland.constant.TYPE_REPLY
import com.yanrou.dawnisland.reply.ReplyDialog
import timber.log.Timber
import java.util.*


/**
 * @author suche
 */
class SeriesContentActivity : AppCompatActivity() {
    private lateinit var recyclerView: RecyclerView
    private lateinit var toolbar: Toolbar
    private lateinit var actionBar: ActionBar
    var id: String? = null
    private var forumName: String? = null
    private lateinit var smartRefreshLayout: SmartRefreshLayout
    private var multiTypeAdapter: MultiTypeAdapter? = null
    private lateinit var viewModel: SeriesContentViewModel
    private val jumpPopup: JumpPopup by lazy { JumpPopup(this) }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.series_content_menu, menu)
        val onlyPoSwitch = menu.findItem(R.id.only_po_switch)
        viewModel.onlyPoLiveData.observe(this, Observer { t ->
            if (t) {
                onlyPoSwitch.title = "查看所有"
            } else {
                onlyPoSwitch.title = "只看po"
            }
        })
        menu.findItem(R.id.subscribe).setOnMenuItemClickListener {
            val subscriptionId = PreferenceManager.getDefaultSharedPreferences(baseContext).getString("subscriber_id", "666")
            viewModel.addFeed(subscriptionId!!, id!!)
            Toast.makeText(baseContext, "添加订阅", Toast.LENGTH_SHORT).show()
            true
        }
        return true
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_series_content)
        //获取控件
        initView()
        actionBar.setDisplayHomeAsUpEnabled(true)
        val intent = intent
        id = intent.getStringExtra("id")
        forumName = intent.getStringExtra("forumTextView")
        toolbar.title = "A岛 · $forumName"
        toolbar.subtitle = ">>No.$id · adnmb.com"
        viewModel = ViewModelProvider(this).get(SeriesContentViewModel::class.java)
        viewModel.referenceHandler = {
            QuotePopup(this).showQuote(it, "")
        }
        viewModel.seriesId = id.toString()
        multiTypeAdapter = MultiTypeAdapter()
        multiTypeAdapter!!.register(ContentItem::class.java, ContentViewBinder())
        val layoutManager = LinearLayoutManager(this)
        recyclerView.layoutManager = layoutManager
        //用于刷新的监听器
        recyclerView.addOnScrollListener(object : SeriesRecyclerOnScrollListener() {
            override fun onLoadMore() {
                viewModel.loadMore()
            }
        })
        //用于报告页数的监听器
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                viewModel.nowIndex = layoutManager.findLastVisibleItemPosition()
            }
        })
        recyclerView.adapter = multiTypeAdapter
        smartRefreshLayout.setEnableAutoLoadMore(false)
        smartRefreshLayout.setOnRefreshListener {
            viewModel.loadPreviousPage(layoutManager.findLastVisibleItemPosition())
        }

        smartRefreshLayout.setOnLoadMoreListener { viewModel.loadMore() }
        viewModel.listLiveData.observe(this, Observer { contentItems ->
            val oldList = multiTypeAdapter!!.items
            //创建一个新的表
            val newList: List<Any> = ArrayList<Any>(contentItems!!)
            val diffResult = DiffUtil.calculateDiff(DiffContentList(oldList, newList))
            multiTypeAdapter!!.items = newList

            diffResult.dispatchUpdatesTo(multiTypeAdapter!!)
            smartRefreshLayout.finishRefresh()
            smartRefreshLayout.finishLoadMore()


        })
        //告诉view model可以开始读取数据了
        viewModel.firstStart()
    }

    @Suppress("unused")
    private fun makeStatusBarTran(window: Window) {
        window.clearFlags(WindowManager.LayoutParams.FLAG_TRANSLUCENT_STATUS or WindowManager.LayoutParams.FLAG_TRANSLUCENT_NAVIGATION)
        window.addFlags(WindowManager.LayoutParams.FLAG_DRAWS_SYSTEM_BAR_BACKGROUNDS)
        window.statusBarColor = ContextCompat.getColor(this, R.color.colorPrimary)
    }

    private fun initView() {
        recyclerView = findViewById(R.id.series_content_recycleview)
        toolbar = findViewById(R.id.toolbar)
        setSupportActionBar(toolbar)
        actionBar = this.supportActionBar!!
        smartRefreshLayout = findViewById(R.id.refresher)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
            R.id.copy_series_id -> {
                copyToClipboard(this, ">>No." + viewModel.seriesId)
            }
            R.id.jump_page -> {
                val page = viewModel.getNowPage((recyclerView.layoutManager as LinearLayoutManager)
                        .findLastCompletelyVisibleItemPosition()) ?: 1
                XPopup.Builder(baseContext)
                        .setPopupCallback(object : SimpleCallback() {
                            override fun beforeShow() {
                                super.beforeShow()
                                jumpPopup.updatePages(page, viewModel.maxPage)
                            }
                        })
                        .asCustom(jumpPopup)
                        .show()
                        .dismissWith {
                            if (jumpPopup.submit) {
                                Timber.i("jumping to ${jumpPopup.targetPage}")
                                viewModel.jumpPage(jumpPopup.targetPage)
                            }
                        }
            }
            R.id.reply -> {
                val replyDialog = ReplyDialog()
                val bundle = Bundle()
                bundle.putString("seriesId", id)
                bundle.putInt(TYPE_KEY, TYPE_REPLY)
                replyDialog.arguments = bundle
                val ft = supportFragmentManager.beginTransaction()
                if (supportFragmentManager.findFragmentByTag("reply") == null) {
                    ft.add(replyDialog, "reply")
                    ft.commit()
                }
            }
            R.id.only_po_switch -> {
                viewModel.switchToOnlyPo()
            }
            else -> {
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun copyToClipboard(context: Context, content: CharSequence?) {
        val clipboard: ClipboardManager = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        clipboard.setPrimaryClip(ClipData.newPlainText(null, content)) //参数一：标签，可为空，参数二：要复制到剪贴板的文本
    }


}